cmd/enterprise-portal/internal/dotcomdb/dotcomdb.go

```diff
@@ -1,32 +1,250 @@
+		limits.EmbeddingsSource = codyaccessv1.CodyGatewayRateLimitSource_CODY_GATEWAY_RATE_LIMIT_SOURCE_OVERRIDE
+		limits.Embeddings.IntervalSeconds = *c.EmbeddingsRateSeconds
+	}
+
+	return limits
+}
+
+func (c CodyGatewayAccessAttributes) GenerateAccessTokens() []string {
+	accessTokens := make([]string, len(c.LicenseKeyHashes))
+	for i, t := range c.LicenseKeyHashes {
+		// See license.GenerateLicenseKeyBasedAccessToken
+		accessTokens[i] = license.LicenseKeyBasedAccessTokenPrefix + hex.EncodeToString(t)
+	}
+	return accessTokens
+}
+
+var ErrCodyGatewayAccessNotFound = errors.New("cody gateway access not found")
+
+const selectCodyGatewayAccessAttributes = `
+SELECT
+	ps.id,
+	ps.cody_gateway_enabled,
+	-- ChatCompletionsRateLimit override
+	ps.cody_gateway_chat_rate_limit,
+	ps.cody_gateway_chat_rate_interval_seconds,
+	-- CodeCompletionsRateLimit override
+	ps.cody_gateway_code_rate_limit,
+	ps.cody_gateway_code_rate_interval_seconds,
+	-- EmbeddingsRateLimit override
+	ps.cody_gateway_embeddings_api_rate_limit,
+	ps.cody_gateway_embeddings_api_rate_interval_seconds,
+	-- "Active license"
+	MAX(active_license.license_tags),
+	MAX(active_license.license_user_count),
+	-- All past license keys that can be used as "access tokens"
+	array_agg(tokens.license_key_hash) as license_key_hashes
+FROM product_subscriptions ps
+	LEFT JOIN (
+		SELECT DISTINCT ON (product_subscription_id)
+			product_subscription_id,
+			license_tags,
+			license_user_count
+		FROM product_licenses
+		-- Get most recently created license key as the "active license"
+		ORDER BY product_subscription_id, created_at DESC
+	) active_license ON active_license.product_subscription_id = ps.id
+	LEFT JOIN (
+		SELECT
+			product_subscription_id,
+			digest(license_key, 'sha256') as license_key_hash
+		FROM product_licenses
+		WHERE access_token_enabled IS TRUE
+	) tokens ON tokens.product_subscription_id = ps.id
+`
+
+type GetCodyGatewayAccessAttributesOpts struct {
+	BySubscription *string
+	ByAccessToken  *string
+}
+
+func (r *Reader) GetCodyGatewayAccessAttributesBySubscription(ctx context.Context, subscriptionID string) (*CodyGatewayAccessAttributes, error) {
+	query := selectCodyGatewayAccessAttributes +
+		"WHERE ps.id = $1 GROUP BY ps.id"
+	row := r.conn.QueryRow(ctx, query, subscriptionID)
+	return r.scanOneCodyGatewayAccessAttributes(row)
+}
+
+func (r *Reader) GetCodyGatewayAccessAttributesByAccessToken(ctx context.Context, token string) (*CodyGatewayAccessAttributes, error) {
+	// Below is copy-pasta from 'func (t dbTokens) LookupProductSubscriptionIDByAccessToken'
+	// in 'cmd/frontend/internal/dotcom/productsubscription'.
+	if !strings.HasPrefix(token, productsubscription.AccessTokenPrefix) &&
+		!strings.HasPrefix(token, license.LicenseKeyBasedAccessTokenPrefix) {
+		return nil, errors.WithSafeDetails(ErrCodyGatewayAccessNotFound, "invalid token with unknown prefix")
+	}
+	decoded, err := hex.DecodeString(token[len(license.LicenseKeyBasedAccessTokenPrefix):])
```

Comment:
For readability, consider moving the non-trivial expression to a local variable.
```suggestion
	tokenSansPrefix := token[len(license.LicenseKeyBasedAccessTokenPrefix):]
	decoded, err := hex.DecodeString(tokenSansPrefix)
```